src/happy.rs

```diff
@@ -0,0 +1,134 @@
+use std::io;
+use std::iter::FusedIterator;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::sync::mpsc::{channel, RecvTimeoutError};
+use std::thread;
+use std::time::{Duration, Instant};
+
+const DEFAULT_CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+const RACE_DELAY: Duration = Duration::from_millis(200);
+
+/// This function implements a basic form of the happy eyeballs RFC to quickly connect
+/// to a domain which is available in both IPv4 and IPv6. Connection attempts are raced
+/// against each other and the first to connect successfully wins the race.
+///
+/// If the timeout is not provided, a default timeout of 10 seconds is used.
+pub fn connect<A>(addrs: A, timeout: impl Into<Option<Duration>>) -> io::Result<TcpStream>
+where
+    A: ToSocketAddrs,
+{
+    let timeout = timeout.into().unwrap_or(DEFAULT_CONNECTION_TIMEOUT);
+    let addrs: Vec<_> = addrs.to_socket_addrs()?.collect();
+
+    if addrs.len() == 1 {
```

Comment:
I think this could be a slice pattern like `if let [addr] = addrs.as_slice()` to check and unpack in one go.